### ISSUE TITLE
[REGRESSION] Wrong `systemd-analyze security` output with regard to `SystemCallFilter=`

### ISSUE THREAD

**Issue by pfactum** [systemd/systemd#23663]
**systemd version the issue has been seen with**

```
systemd 251 (251.2-1-arch)
+PAM +AUDIT -SELINUX -APPARMOR -IMA +SMACK +SECCOMP +GCRYPT +GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 -PWQUALITY +P11KIT -QRENCODE +TPM2 +BZIP2 +LZ4 +XZ +ZLIB +ZSTD -BPF_FRAMEWORK +XKBCOMMON +UTMP -SYSVINIT default-hierarchy=unified
```

**Used distribution**

Arch Linux

**Linux kernel version used** (`uname -a`)

```
[~]> uname -a
Linux spock 5.18.0-pf2 #1 SMP PREEMPT_DYNAMIC Mon, 06 Jun 2022 14:51:13 +0000 x86_64 GNU/Linux
```

**CPU architecture issue was seen on**

`x86_64`

**Expected behaviour you didn't see**

Having multiple `SystemCallFilter=` lines in the `.service` file I expect to see that the first one defines default behaviour (_allowlisting vs denylisting_) as explained in the documentation:

> If you specify both types of this option (i.e. allow-listing and deny-listing), the first encountered will take precedence and will dictate the default action (termination or approval of a system call). Then the next occurrences of this option will add or delete the listed system calls from the set of the filtered system calls, depending of its type and the default action.

This is not the case however.

**Unexpected behaviour you saw**

I've got the following lines in my `.service` file:

```
[~]> grep -n SystemCallFilter= /usr/lib/systemd/system/tinc@.service
33:SystemCallFilter=@system-service
34:SystemCallFilter=~@resources @privileged
```

And this is what `systemd-analyze security` gives me:

```
[~]> systemd-analyze security --offline=true /usr/lib/systemd/system/tinc@.service | grep SystemCallFilter=
✗ SystemCallFilter=~@swap                                     System call allow list defined for service, and @swap is included (e.g. swapoff is allowed)                  0.2
✗ SystemCallFilter=~@resources                                System call allow list defined for service, and @resources is included (e.g. ioprio_set is allowed)          0.2
✗ SystemCallFilter=~@reboot                                   System call allow list defined for service, and @reboot is included (e.g. kexec_file_load is allowed)        0.2
✗ SystemCallFilter=~@raw-io                                   System call allow list defined for service, and @raw-io is included (e.g. ioperm is allowed)                 0.2
✗ SystemCallFilter=~@privileged                               System call allow list defined for service, and @privileged is included (e.g. chown is allowed)              0.2
✗ SystemCallFilter=~@obsolete                                 System call allow list defined for service, and @obsolete is included (e.g. _sysctl is allowed)              0.1
✗ SystemCallFilter=~@mount                                    System call allow list defined for service, and @mount is included (e.g. chroot is allowed)                  0.2
✗ SystemCallFilter=~@module                                   System call allow list defined for service, and @module is included (e.g. delete_module is allowed)          0.2
✗ SystemCallFilter=~@debug                                    System call allow list defined for service, and @debug is included (e.g. lookup_dcookie is allowed)          0.2
✗ SystemCallFilter=~@cpu-emulation                            System call allow list defined for service, and @cpu-emulation is included (e.g. modify_ldt is allowed)      0.1
✗ SystemCallFilter=~@clock                                    System call allow list defined for service, and @clock is included (e.g. adjtimex is allowed)                0.2
```

This is not correct, because I've excluded at least `@privileged` and `@resources` explicitly.

(_the same applies to online check, so `--offline=true` option doesn't make any difference_)

**Steps to reproduce the problem**

My full `.service` file is shown below:

```
[Unit]
Description=Tinc net %i
Documentation=info:tinc
Documentation=man:tinc(8) man:tinc.conf(5)
Documentation=http://tinc-vpn.org/docs/
PartOf=tinc.service
ReloadPropagatedFrom=tinc.service

[Service]
DynamicUser=true
User=tinc-%i
Group=tinc-%i
AmbientCapabilities=CAP_NET_ADMIN CAP_NET_BIND_SERVICE
CapabilityBoundingSet=CAP_NET_ADMIN CAP_NET_BIND_SERVICE
RestrictAddressFamilies=AF_INET AF_INET6 AF_UNIX
RestrictNamespaces=true
PrivateDevices=true
BindPaths=/dev/net/tun
DeviceAllow=/dev/net/tun rw
ProtectClock=true
ProtectControlGroups=true
ProtectHome=true
ProtectKernelLogs=true
ProtectKernelModules=true
ProtectKernelTunables=yes
ProtectProc=invisible
ProcSubset=pid
SystemCallArchitectures=native
RestrictRealtime=true
LockPersonality=true
MemoryDenyWriteExecute=true
ProtectHostname=true
SystemCallFilter=@system-service
SystemCallFilter=~@resources @privileged
TasksMax=2
UMask=0066
WorkingDirectory=/var/lib/tinc/%i
ConfigurationDirectory=tinc/%i
RuntimeDirectory=tinc/%i
StateDirectory=tinc/%i
PIDFile=${RUNTIME_DIRECTORY}/pid
LoadCredential=%i:/etc/credstore/tinc/%i
ExecStart=/usr/bin/tincd -D --pidfile=${RUNTIME_DIRECTORY}/pid -n %i -o Ed25519PrivateKeyFile=${CREDENTIALS_DIRECTORY}/%i_ed25519_key.priv -o PrivateKeyFile=${CREDENTIALS_DIRECTORY}/%i_rsa_key.priv -o Interface=tinc-%i
ExecReload=/usr/bin/tinc --pidfile=${RUNTIME_DIRECTORY}/pid -n %i reload
Type=notify
WatchdogSec=10
Restart=on-failure

[Install]
WantedBy=tinc.service
```

**Additional program output to the terminal or log subsystem illustrating the issue**

It was not the case in the past. I cannot say when the issue was introduced, but IIRC v250 behaved correctly.

Also, if I swap lines in the `.service` file so that the denylist comes first, then `systemd-analyze security` doesn't report any issue:

```
[~]> grep -n SystemCallFilter= /usr/lib/systemd/system/tinc@.service
33:SystemCallFilter=~@resources @privileged
34:SystemCallFilter=@system-service

[~]> systemd-analyze security --offline=true /usr/lib/systemd/system/tinc@.service | grep SystemCallFilter=
✓ SystemCallFilter=~@swap                                     System call deny list defined for service, and @swap is included
✓ SystemCallFilter=~@resources                                System call deny list defined for service, and @resources is included
✓ SystemCallFilter=~@reboot                                   System call deny list defined for service, and @reboot is included
✓ SystemCallFilter=~@raw-io                                   System call deny list defined for service, and @raw-io is included
✓ SystemCallFilter=~@privileged                               System call deny list defined for service, and @privileged is included
✓ SystemCallFilter=~@obsolete                                 System call deny list defined for service, and @obsolete is included
✓ SystemCallFilter=~@mount                                    System call deny list defined for service, and @mount is included
✓ SystemCallFilter=~@module                                   System call deny list defined for service, and @module is included
✓ SystemCallFilter=~@debug                                    System call deny list defined for service, and @debug is included
✓ SystemCallFilter=~@cpu-emulation                            System call deny list defined for service, and @cpu-emulation is included
✓ SystemCallFilter=~@clock                                    System call deny list defined for service, and @clock is included
```

But this goes against the documentation.

**Comment by adrian5** [comment]
@yuwata Can we assume this is a regression in `systemd-analyze` only, not in functionality of the actual unit?

**Comment by yuwata** [comment]
Not sure, but I hope so.

**Comment by yuwata** [comment]
Fix is waiting in #23756. Fortunately, this is just an issue in systemd-analyze.
